neqo-transport/src/qlog.rs

```diff
@@ -0,0 +1,409 @@
+    if let Some(qlog) = qlog {
+        let mut qlog = qlog.borrow_mut();
+        let elapsed = now.duration_since(qlog.zero_time);
+
+        let mut frames = Vec::new();
+        let mut d = Decoder::from(body);
+
+        while d.remaining() > 0 {
+            match Frame::decode(&mut d) {
+                Ok(f) => frames.push(frame_to_qlogframe(&f)),
+                Err(_) => {
+                    qinfo!("qlog: invalid frame");
+                    break;
+                }
+            }
+        }
+
+        let packet_type = pkt_type_to_qlog_pkt_type(pt);
+
+        qlog.trace.push_transport_event(
+            format!("{}", elapsed.as_micros()),
+            TransportEventType::PacketSent,
+            EventData::PacketSent {
+                packet_type,
+                header: PacketHeader {
+                    packet_number: pn.to_string(),
+                    packet_size: None,
+                    payload_length: None,
+                    version: None,
+                    scil: None,
+                    dcil: None,
+                    scid: None,
+                    dcid: None,
+                },
+                frames: Some(frames),
+                is_coalesced: None,
+                raw_encrypted: None,
+                raw_decrypted: None,
+            },
+        )
+    }
+}
+
+pub fn packet_received(qlog: &Option<NeqoQlogRef>, now: Instant, payload: &DecryptedPacket) {
+    if let Some(qlog) = qlog {
+        let mut qlog = qlog.borrow_mut();
+        let elapsed = now.duration_since(qlog.zero_time);
+
+        let mut frames = Vec::new();
+        let mut d = Decoder::from(&payload[..]);
+
+        while d.remaining() > 0 {
+            match Frame::decode(&mut d) {
+                Ok(f) => frames.push(frame_to_qlogframe(&f)),
+                Err(_) => {
+                    qinfo!("qlog: invalid frame");
+                    break;
+                }
+            }
+        }
+
+        let packet_type = pkt_type_to_qlog_pkt_type(payload.packet_type());
+
+        qlog.trace.push_transport_event(
+            format!("{}", elapsed.as_micros()),
+            TransportEventType::PacketReceived,
+            EventData::PacketReceived {
+                packet_type,
+                header: PacketHeader {
+                    packet_number: payload.pn().to_string(),
+                    packet_size: None,
+                    payload_length: None,
+                    version: None,
+                    scil: None,
+                    dcil: None,
+                    scid: None,
+                    dcid: None,
+                },
+                frames: Some(frames),
+                is_coalesced: None,
+                raw_encrypted: None,
+                raw_decrypted: None,
+            },
+        )
+    }
+}
+
+// Helper functions
+
+fn frame_to_qlogframe(frame: &Frame) -> QuicFrame {
+    match frame {
+        Frame::Padding => QuicFrame::Padding {
+            frame_type: QuicFrameTypeName::Padding,
```

Comment:
Hang on...  Why do we even need to populate this field?  Why can't the qlog crate do that for us?  After all, it duplicates the enum type.

Reply:
Yay, qlog 0.2 does this now!